drivers: gpio: gpio_mcux_lpc: add support for module interrupts, and implement PINT interrupt controller driver

danieldegrasse opened the pull request:
LPC gpio driver has historically used the PINT to route interrupts, which limits interrupt routing to GPIO ports that the PINT supports as inputs. Add the ability to configure the interrupt source for each GPIO driver, so the user can allocate the two interrupts shared between all ports as needed for their application.

Note that this still subject to the limitation in Zephyr that device instances cannot share interrupts. Thus, only two additional GPIO ports can have interrupts enabled using this feature.


This PR now also includes a interrupt controller driver for NXP's PINT (pin interrupt and pattern match engine), in order to remove the integration of this IP block directly into the LPC GPIO driver

decsny commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -357,6 +373,87 @@ static int gpio_mcux_lpc_pin_interrupt_configure(const struct device *dev,
+	/* Route interrupt to source A or B based on interrupt source */
+	int_idx = int_source == INT_SOURCE_INTA ? kGPIO_InterruptA : kGPIO_InterruptB;
+
+	switch (mode) {
```
```
        /* Disable interrupt if requested */
	if(mode ==  GPIO_INT_MODE_DISABLED){
	        GPIO_PinDisableInterrupt(config->gpio_base, config->port_no, pin, int_idx);
		return 0;
	}
	
        /* Set pin interrupt level */
	if(mode == GPIO_INT_MODE_LEVEL) {
	        pin_config.mode = kGPIO_PinIntEnableLevel;
	}
	else if (mode ==  GPIO_INT_MODE_EDGE) {
	       pin_config.mode = kGPIO_PinIntEnableEdge;
	}
	else {
	        return -ENOTSUP;
	}
	
        /* Set pin interrupt trigger */
	if (trig == GPIO_INT_TRIG_HIGH) {
		pin_config.polarity = kGPIO_PinIntEnableHighOrRise;
	} else if (trig == GPIO_INT_TRIG_LOW) {
		pin_config.polarity = kGPIO_PinIntEnableLowOrFall;
	} else {
		return -ENOTSUP;
	}
	
        /* Enable interrupt with new configuration */
	GPIO_SetPinInterruptConfig(config->gpio_base, config->port_no, pin, &pin_config);
	GPIO_PinEnableInterrupt(config->gpio_base, config->port_no, pin, int_idx);
	return 0;
```

I don't like the switch statement

danieldegrasse replied:
Sure, I'll add this

danieldegrasse commented on dts/bindings/gpio/nxp,lpc-gpio.yaml:
```diff
@@ -28,6 +28,20 @@ properties:
 
+  int-source:
+    type: string
+    default: "pint"
```
I would prefer to leave pint as the default. It seems all the LPC devices use it as an interrupt source (which is why this driver was using PINT to manage interrupts previously). For the RT devices, if PINT is the default then interrupt installation will simply fail on GPIO ports that cannot use PINT as a source, until they are configured otherwise.

decsny replied:
Considering that PINT is not part of the GPIO IP and it is a separate IP block only present on some SOCs, it does not feel right to say the default hardware configuration for this GPIO IP block uses PINT. I would even go so far as to say that PINT should have it's own DT node. But since it is based on using the GPIO pins, it might make sense to make it a child node of the GPIO or vice versa, or to make a phandle property for pint. I just think it needs to be clear in devicetree somehow that PINT is not part of the LPC GPIO IP block. And I think PINT could have an interrupt_controller driver similar to how STM32 has an [EXTI driver](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/interrupt_controller/intc_exti_stm32.c) for dealing with external interrupt sources, which seems to be a very similar concept to what PINT is.

henrikbrixandersen replied:
> And I think PINT could have an interrupt_controller driver similar to how STM32 has an [EXTI driver](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/interrupt_controller/intc_exti_stm32.c) for dealing with external interrupt sources, which seems to be a very similar concept to what PINT is.

I agree. The PINT needs to be represented in DTS. Comparing it to the STM32 EXTI seems accurate.

henrikbrixandersen reviewed (changes requested):
Please convert this to representing the PINT as DTS node, representing an interrupt controller. The STM32 EXTI can be used as inspiration.

danieldegrasse committed:
drivers: interrupt_controller: introduce PINT driver

Introduce PINT driver, for NXP pin interrupt and pattern match engine.
The driver currently supports only the pin interrupt feature of the
PINT.

Add DTS entires for the PINT on LPC and RT devices that support this
peripheral, and remove the interrupt defintions that are PINT specific
from the GPIO module on these devices.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: gpio: gpio_mcux_lpc: remove PINT code

Remove PINT management code from the LPC GPIO driver, as this code is
now contained within the PINT interrupt controller driver, which exposes
an interface to install interrupt callbacks.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: gpio: gpio_mcux_lpc: add support for module interrupts

On iMX.RT devices, the number of GPIO pins exceeds the maximum of
64 that the PINT interrupt controller can support. Therefore, two
interrupt lines are now shared between the GPIO modules.

This patch allows the user to set the interrupt source for a GPIO
peripheral. For most LPC devices, this will always be the PINT. For some
RT devices, the PINT cannot use pins on GPIO modules other than 0 and 1
as input, and thus the INTA and INTB sources should be used.

Since Zephyr does not support sharing these interrupt between all GPIO
controllers, the user must configure a subset of all GPIO controllers to
use the shared module interrupts. An example of how to do so is provided
for the RT595 EVK.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@henrikbrixandersen Updated PR to convert the PINT to use an interrupt controller driver, and integrated the GPIO module IRQ support on top of that update

decsny commented on drivers/interrupt_controller/intc_nxp_pint.c:
```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2023, NXP
```
no comma

danieldegrasse replied:
Fixed

decsny commented on drivers/interrupt_controller/intc_nxp_pint.c:
```diff
@@ -0,0 +1,204 @@
+	uint8_t slot;
+
+	if (pin > ARRAY_SIZE(pin_pint_id)) {
+		return;
```
nit but why not declare these functions to return int and return -EINVAL since that is essentially what you are checking for here

danieldegrasse replied:
Good question- I considered this, but I don't think we need to alert the user via a return value. In all the cases where the `disable` or `unset` functions returns prematurely, the action that the user requested is effectively not needed, since the pin is either invalid, or doesn't currently an interrupt enabled/callback installed. Therefore, the user doesn't really care *how* the pin's interrupt was disabled, since it still effectively will be.

If the user tries to use an invalid pin, they should find out about that in the `enable` function. I don't think the `disable` function needs to tell them that via a return code, since it shouldn't be called first anyways.

cfriedt reviewed (approved):
Looks like quite an improvement 👍

dleach02 reviewed (approved):
@henrikbrixandersen, can you review your block to ensure Daniel addressed your concerns

henrikbrixandersen reviewed (approved):
Looks good, thank you.
